group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupStore.java

```diff
@@ -0,0 +1,307 @@
+            boolean stateCheck = statesFilter.isEmpty() || group.isInStates(caseInsensitiveFilterSet, committedOffset);
+            boolean typeCheck = enumTypesFilter.isEmpty() || enumTypesFilter.contains(group.type());
+
+            return stateCheck && typeCheck;
+        };
+
+        Stream<Group> groupStream = groups.values(committedOffset).stream();
+
+        return groupStream
+            .filter(combinedFilter)
+            .map(group -> group.asListedGroup(committedOffset))
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * @return The group corresponding to the group id or throw GroupIdNotFoundException.
+     */
+    public Group group(String groupId) throws GroupIdNotFoundException {
+        Group group = groups.get(groupId, Long.MAX_VALUE);
+        if (group == null) {
+            throw new GroupIdNotFoundException(String.format("Group %s not found.", groupId));
+        }
+        return group;
+    }
+
+    /**
+     * @return The group corresponding to the group id at the given committed offset
+     *         or throw GroupIdNotFoundException.
+     */
+    public Group group(String groupId, long committedOffset) throws GroupIdNotFoundException {
+        Group group = groups.get(groupId, committedOffset);
+        if (group == null) {
+            throw new GroupIdNotFoundException(String.format("Group %s not found.", groupId));
+        }
+        return group;
+    }
+
+    /**
+     * @return The set of groups subscribed to the topic.
+     */
+    public Set<String> groupsSubscribedToTopic(String topicName) {
+        Set<String> groups = groupsByTopics.get(topicName);
+        return groups != null ? groups : Collections.emptySet();
+    }
+
+    /**
+     * A new metadata image is available.
+     *
+     * @param newImage  The new metadata image.
+     * @param delta     The delta image.
+     */
+    public void onNewMetadataImage(MetadataImage newImage, MetadataDelta delta) {
+        metadataImage = newImage;
+
+        // Notify all the groups subscribed to the created, updated or
+        // deleted topics.
+        Optional.ofNullable(delta.topicsDelta()).ifPresent(topicsDelta -> {
+            Set<String> allGroupIds = new HashSet<>();
+            topicsDelta.changedTopics().forEach((topicId, topicDelta) -> {
+                String topicName = topicDelta.name();
+                allGroupIds.addAll(groupsSubscribedToTopic(topicName));
+            });
+            topicsDelta.deletedTopicIds().forEach(topicId -> {
+                TopicImage topicImage = delta.image().topics().getTopic(topicId);
+                allGroupIds.addAll(groupsSubscribedToTopic(topicImage.name()));
+            });
+            allGroupIds.forEach(groupId -> {
+                Group group = groups.get(groupId);
+                if (group != null && (group.type() == CONSUMER || group.type() == SHARE)) {
+                    ((ModernGroup<?>) group).requestMetadataRefresh();
```

Comment:
I don't really like this. We should just do proper object orientation here, define `oneMetadataUpdate` or something like that to `Group` and call it here on any group type, and leave it empty for classic groups.

But I see that this is not your code, but just copied. So, we could fix it in a later PR.